pkg/testcase/certificates.go

```diff
@@ -0,0 +1,122 @@
+package testcase
+
+import (
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/rancher/distros-test-framework/factory"
+	"github.com/rancher/distros-test-framework/shared"
+)
+
+func errCheck(error error, message string, ignore bool) {
+	if error != nil {
+		shared.LogLevel("error", message)
+		if !ignore {
+			Expect(error).NotTo(HaveOccurred())
+		}
+
+	}
+}
+
+// Rotate certificate for etcd only and cp only nodes
+func certRotate(product string, ip string) {
+	// Stop service on server
+	_, stopError := shared.StopService(product, ip, "server")
+	errCheck(stopError, fmt.Sprintf("Error stopping %s service", product), false)
+	time.Sleep(time.Duration(10) * time.Second) // Adding 10 second sleep between steps
+	// Rotate certificate
+	_, rotateError := shared.CertRotate(product, ip)
+	errCheck(rotateError, fmt.Sprintf("Error running certificate rotate for %s service", product), false)
+	time.Sleep(time.Duration(10) * time.Second) // Adding 10 second sleep between steps
+	// start service on server
+	_, startError := shared.StartService(product, ip, "server")
+	errCheck(startError, fmt.Sprintf("Error starting %s service", product), true)
+	time.Sleep(time.Duration(10) * time.Second) // Adding 10 second sleep between steps
+	// Retry Stop start of service. (Inconsistent failure to start service above,
+	// can be fixed hopefully with a restart)
+	if startError != nil {
+		_, retryError := shared.StopStartService(product, ip, "server")
+		errCheck(retryError, fmt.Sprintf("ERROR: Retrying Stop/Start %s service", product), false)
+	}
+}
+
+// Compare TLS Directories before and after cert rotation to display identical files
+func compareTLSDir(product string, ip string) (string, error) {
+	dataDir := fmt.Sprintf("/var/lib/rancher/%s", product)
+	serverDir := fmt.Sprintf("%s/server", dataDir)
+	origTLSDir := fmt.Sprintf("%s/tls", serverDir)
+	cmd := fmt.Sprintf("sudo ls -lt %s/ | grep tls | awk {'print $9'} | sed -n '2 p'", serverDir)
+	tlsDir, error := shared.RunCommandOnNode(cmd, ip)
+	if error != nil {
+		shared.LogLevel("warn", "Unable to get new TLS Directory name")
+	}
+	shared.LogLevel("info", fmt.Sprintf("TLS Directory name: %s", tlsDir))
+	newTLSDir := fmt.Sprintf("%s/%s", serverDir, tlsDir)
+	shared.LogLevel("info", "Comparing Directories: %s and %s", origTLSDir, newTLSDir)
+	cmd2 := fmt.Sprintf("sudo diff -sr %s/ %s/ | grep -i identical | "+
+		"awk '{print $2}' | xargs basename -a | "+
+		"awk 'BEGIN{print \"Identical Files:  \"}; {print $1}'", origTLSDir, newTLSDir)
+	return shared.RunCommandOnNode(cmd2, ip)
+}
+
+func TestCertRotate() {
+	// Node Types needed for this test:
+	// Server1 - Etcd only node
+	// Server2 - Control Plane only node
+	// Agent1 - Worker node
+	// Initialize the cluster
+	cluster := factory.AddCluster(GinkgoT())
+	// Get server1, server2, product info from cluster
+	server1 := cluster.ServerIPs[0]
+	server2 := cluster.ServerIPs[1]
```

Comment:
product, err := shared.GetProduct()
	ips := shared.FetchNodeExternalIP()
	ipagent := ips[1:]
	ipservers := ips[:1]
	// some logic to split agent and servers ips because here it will be better get from the current cluster state,
	// //not from the immutable data from the addcluster

Reply:
just a tip to try not use the ones came from  cluster := factory.AddCluster(GinkgoT())

Reply:
this did not work. the ips returned are never in the same order - hard to tell which is server/agent. 
going back to the previous way of getting ips for now.

Reply:
my example was just an example actually , not an implementation for you ! lol
 thats why i added the comment 
// some logic to split agent and servers ips because here it will be better get from the current cluster state,

if you want a tip, u can create 2 preivous lists with 	 cluster.ServerIP  cluster.AgentIPs , then get one more list from  FetchNodeExternalIP , then compare ips,  then split into 2 newlists  ( finalAgents , finalServers ) 

but tbh, since u are not rebooting nor even deleting,  its not that bad use those that u are using, its just to avoid use globals